interp/iostream_test.go

```diff
@@ -0,0 +1,321 @@
+package interp
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"math/rand"
+	"os"
+	"os/exec"
+	"runtime"
+	"strings"
+	"testing"
+	"time"
+)
+
+const (
+	shellWaitTime          = 500 * time.Millisecond
+	maxRandExeAttempts     = 10
+	outputStreamBufferSize = 1024
+)
+
+type shellTest struct {
+	Name, shell string
+	exitCode    int
+	input       string
+	output      outputChecker
+	requires    string
+	skipWin     bool
+}
+
+func TestStreamInFile(t *testing.T) {
+	t.Run("DoubleClose", func(t *testing.T) {
```

Comment:
Well, these are unit tests for the streams themselves. Part of the advantages of the abstracting them is that we can verify the interface between them and the rest of `interp` is consistent.

Yes, right now it's _presumably_ impossible to double close but what if a bug is introduced that causes double-close to happen? Don't you want to know that the expected error conditions are consistent so that it is visible to the rest of the program?

I think the greatest advantage of these lower-level tests are to check that behavior is consistent for all four types of streams. The is running basically the same test for each type of stream which is why I feel a sub-test is appropriate. Sure a flat structure is simpler, but only because it's less work for the writer. I've already taken the trouble to separate them. There is no advantage in mushing them together now.

Coincidentally, a few days ago I added local tests to check the double close for the command streams as well. The `outCmdStream` does not behave as expected and it does not return an error if it is double-closed. If I didn't actually test the error checking I would not have known that?

Maybe if `iostreams` was a separate package you would feel more comfortable unit testing it?

Reply:
I'm okay leaving a couple of specific tests like the double-close one if you think it's going to catch bugs. However, in the case of the double-close, I'm still not sure it's worth it: why do we need to assert that it returns an error if double-closed when we're not double-closing it in the code?

In any case, I strongly believe we should test as much as possible using interp or similar white-box tests. These have proved very valuable in the past, for a couple of reasons:

1) I've been able to do large refactorings of the interpreter without the test suite changing one bit. As a result, I have a ton of confidence that the refactorings haven't broken anything. On the flip hand, if the tests are tied to the implementation, you're testing that specific implementation, so when the implementation changes, the tests break (and you don't know if that's because the old implementation was bad, or just because the tests relied on implementation details -- probably the latter). So white-box tests are resilient in the face of implementation changes.

2) All the interp tests are also run through Gawk. This ensures I've got them right and that Gawk has the same behaviour. A small number of tests use `# !gawk` type comments to not run on gawk, but almost all of them are run on both Gawk and goawk. The has helped me numerous times in the past. Similarly, I run the entire nawk and gawk test suites on Goawk as well, and this is of course done white-box, by feeding source code to GoAWK, sending input, and checking output.

Once again, I'm okay with leaving a couple of specific "programmer error" tests in `iostream_test.go`, but let's redo the rest as interp tests. (I don't think we should make `iostreams` a separate package.)

Reply:
Oops, I said "white box" above when I meant "black box". Updating.

Reply:
I've left a single `TestStreamDoubleClose` test inside of `iostream_test.go` and it has the 4 combinations of input/output file/pipe as subtests.